storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -21,70 +21,211 @@
+		}
+
+		unmergedSeries, ok := unmergedShard[series.ID]
 		if !ok {
-			blocks = block.NewDatabaseSeriesBlocks(bopts.GetDatabaseBlockOptions())
-			seriesMap.AddSeries(series.ID, blocks)
+			unmergedSeries = make(map[time.Time][]encoder)
+			unmergedShard[series.ID] = unmergedSeries
 		}
 
-		block := blocks.GetBlockOrAdd(dp.Timestamp.Truncate(blockSize))
-		if err := block.Write(dp.Timestamp, dp.Value, unit, annotation); err != nil {
+		var (
+			err           error
+			unmergedBlock = unmergedSeries[blockStart]
+			wroteExisting = false
+		)
+		for i := range unmergedBlock {
+			if unmergedBlock[i].lastWriteAt.Before(dp.Timestamp) {
+				unmergedBlock[i].lastWriteAt = dp.Timestamp
+				err = unmergedBlock[i].enc.Encode(dp, unit, annotation)
+				wroteExisting = true
+				break
+			}
+		}
+		if !wroteExisting {
+			enc := encoderPool.Get()
+			enc.Reset(blockStart, blopts.GetDatabaseBlockAllocSize())
+
+			err = enc.Encode(dp, unit, annotation)
+			if err == nil {
+				unmergedBlock = append(unmergedBlock, encoder{
+					lastWriteAt: dp.Timestamp,
+					enc:         enc,
+				})
+				unmergedSeries[blockStart] = unmergedBlock
+			}
+		}
+		if err != nil {
 			errs++
 		}
 	}
 	if errs > 0 {
-		log.Errorf("error bootstrapping from commit log: %d block encode errors", errs)
+		s.log.Errorf("error bootstrapping from commit log: %d block encode errors", errs)
 	}
 	if err := iter.Err(); err != nil {
-		log.Errorf("error reading commit log: %v", err)
+		s.log.Errorf("error reading commit log: %v", err)
+	}
+
+	errs = 0
+	emptyErrs := 0
+	result := bootstrap.NewResult()
+	blocksPool := bopts.GetDatabaseBlockOptions().GetDatabaseBlockPool()
+	multiReaderIteratorPool := blopts.GetMultiReaderIteratorPool()
+	for shard, unmergedShard := range unmerged {
+		shardResult := bootstrap.NewShardResult(s.opts.GetBootstrapOptions())
+		for id, unmergedBlocks := range unmergedShard {
+			blocks := block.NewDatabaseSeriesBlocks(blopts)
+			for start, unmergedBlock := range unmergedBlocks {
+				block := blocksPool.Get()
+				if len(unmergedBlock) == 0 {
+					emptyErrs++
+					continue
+				} else if len(unmergedBlock) == 1 {
+					block.Reset(start, unmergedBlock[0].enc)
+				} else {
+					readers := make([]io.Reader, len(unmergedBlock))
+					for i := range unmergedBlock {
+						readers[i] = unmergedBlock[i].enc.Stream()
+					}
+
+					iter := multiReaderIteratorPool.Get()
+					iter.Reset(readers)
+
+					var err error
+					enc := encoderPool.Get()
+					enc.Reset(start, blopts.GetDatabaseBlockAllocSize())
+					for iter.Next() {
+						dp, unit, annotation := iter.Current()
+						encodeErr := enc.Encode(dp, unit, annotation)
+						if encodeErr != nil {
+							err = encodeErr
+							errs++
+							break
+						}
+					}
+					if err != nil {
+						err = iter.Err()
+					}
+
+					iter.Close()
+					for i := range unmergedBlock {
+						unmergedBlock[i].enc.Close()
+					}
+
+					if err != nil {
+						continue
```

Comment:
I see, so we log how many errors happened during the process, which is probably fine